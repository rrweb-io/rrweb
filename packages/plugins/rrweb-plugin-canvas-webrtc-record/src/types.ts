export interface WebRTCDataChannel {
  nodeId: number;
  streamId: string;
}
