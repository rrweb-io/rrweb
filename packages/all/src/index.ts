export * from 'rrweb';
export * from '@rrweb/packer';
// export * from '@rrweb/rrweb-plugin-console-record';
// export * from '@rrweb/rrweb-plugin-console-replay';
