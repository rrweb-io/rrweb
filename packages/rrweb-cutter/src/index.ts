export * from './cut-session';
export * from './prune-branches';
