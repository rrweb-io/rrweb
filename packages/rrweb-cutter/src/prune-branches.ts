import { serializedNodeWithId } from 'rrweb-snapshot';
import type {
  addedNodeMutation,
  eventWithTime,
  mousePosition,
} from '@rrweb/types';
import { IncrementalSource } from 'rrweb';
import { EventType } from 'rrweb';
import { SyncReplayer } from 'rrweb';

export function pruneBranches(
  events: eventWithTime[],
  { keep }: { keep: number[] },
): eventWithTime[] {
  const result: eventWithTime[] = [];
  const replayer = new SyncReplayer(events);
  const treeSet = new Set<number>(keep);
  replayer.reversePlay(({ event }) => {
    if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const tree = getTreeForId(treeSet, node, keep);
      tree.forEach((id) => treeSet.add(id));
    } else if (event.type === EventType.IncrementalSnapshot) {
      if (event.data.source === IncrementalSource.Mutation) {
        const { adds, removes } = event.data;
        removes.forEach((remove) => {
          if (treeSet.has(remove.id)) treeSet.add(remove.parentId);
        });
        adds.forEach((add) => {
          const tree = getTreeForId(treeSet, add.node, keep);
          if (tree.size) {
            treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          } else if (
            'childNodes' in add.node &&
            add.node.childNodes.length > 0
          ) {
            const tree = getTreeForId(treeSet, add.node, keep);
            if (tree.size) treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          }
        });
      }
    }
    return true;
  });

  replayer.play(({ event }) => {
    if (
      [EventType.Meta, EventType.Load, EventType.DomContentLoaded].includes(
        event.type,
      )
    ) {
      result.push(event);
    } else if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const prunedNode = reconstructTreeWithIds(node, treeSet);
      if (prunedNode)
        result.push({
          ...event,
          data: {
            ...event.data,
            node: prunedNode,
          },
        } as eventWithTime);
    } else if (event.type === EventType.IncrementalSnapshot) {
      if ('positions' in event.data) {
        const { positions } = event.data;
        const prunedPositions: mousePosition[] = positions.filter((p) =>
          treeSet.has(p.id),
        );
        if (prunedPositions.length > 0)
          result.push({
            ...event,
            data: {
              ...event.data,
              positions: prunedPositions,
            },
          } as eventWithTime);
      } else if ('id' in event.data) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        if (treeSet.has(event.data.id!)) result.push(event);
      } else if (event.data.source === IncrementalSource.Mutation) {
        const { removes, adds, texts, attributes } = event.data;
        const prunedRemoves = removes.filter((remove) =>
          treeSet.has(remove.id),
        );
        const prunedAdds = adds
          .map((add) =>
            treeSet.has(add.parentId) && keep.includes(add.parentId)
              ? add
              : {
                  ...add,
                  node: reconstructTreeWithIds(add.node, treeSet),
                },
          )
          .filter((add) => Boolean(add.node)) as addedNodeMutation[];
        const prunedTexts = texts.filter((text) => treeSet.has(text.id));
        const prunedAttributes = attributes.filter((attr) =>
          treeSet.has(attr.id),
        );
        if (
          prunedRemoves.length > 0 ||
          prunedAdds.length > 0 ||
          prunedTexts.length > 0 ||
          prunedAttributes.length > 0
        )
          result.push({
            ...event,
            data: {
              ...event.data,
              removes: prunedRemoves,
              adds: prunedAdds,
              texts: prunedTexts,
              attributes: prunedAttributes,
            },
          } as eventWithTime);
      }
    }
    return true;
  });
  return result;
}

export function getTreeForId(
  treeSet: Set<number>,
  node: serializedNodeWithId,
  keepIds: number[],
): Set<number> {
  const results = new Set<number>();
  if (treeSet.has(node.id)) {
    getIdsInNode(node, keepIds).forEach((id) => results.add(id));
  } else if ('childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      const childTree = getTreeForId(treeSet, child, keepIds);
      if (childTree.size > 0) {
        results.add(node.id);
        childTree.forEach((id) => results.add(id));
      }
    }
  }
  return results;
}

export function getIdsInNode(
  node: serializedNodeWithId,
  keepIds: number[],
): Array<number> {
  const results: number[] = [];
  results.push(node.id);
  if (keepIds.includes(node.id) && 'childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      results.push(...getIdsInNode(child, keepIds));
    }
  }
  return results;
}

export function reconstructTreeWithIds(
  node: serializedNodeWithId,
  ids: Set<number>,
): serializedNodeWithId | undefined {
  if (ids.has(node.id)) {
    if ('childNodes' in node) {
      node.childNodes = node.childNodes
        .map((child) => reconstructTreeWithIds(child, ids))
        .filter(Boolean) as serializedNodeWithId[];
    }
    return node;
  }
  return undefined;
}
