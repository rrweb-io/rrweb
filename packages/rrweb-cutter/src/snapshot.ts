import type { serializedNodeWithId, attributes } from 'rrweb-snapshot';
import { elementNode, NodeType, serializedNode } from 'rrweb-snapshot';
import type {
  IRRComment,
  IRRDocument,
  IRRDocumentType,
  IRRElement,
  IRRNode,
  RRCanvasElement,
  RRIFrameElement,
  RRMediaElement,
} from 'rrdom';
import { Mirror } from 'rrdom';

function serializeNode(
  n: IRRNode,
  options: {
    doc: IRRDocument;
    mirror: Mirror;
  },
): serializedNode | false {
  const { doc, mirror } = options;
  // Only record root id when document object is not the base document
  const rootId = getRootId(doc, mirror);
  switch (n.RRNodeType) {
    case NodeType.Document:
      if ((n as IRRDocument).compatMode !== 'CSS1Compat') {
        return {
          type: NodeType.Document,
          childNodes: [],
          compatMode: (n as IRRDocument).compatMode, // probably "BackCompat"
          rootId,
        };
      } else {
        return {
          type: NodeType.Document,
          childNodes: [],
          rootId,
        };
      }
    case NodeType.DocumentType:
      return {
        type: NodeType.DocumentType,
        name: (n as IRRDocumentType).name,
        publicId: (n as IRRDocumentType).publicId,
        systemId: (n as IRRDocumentType).systemId,
        rootId,
      };
    case NodeType.Element:
      return serializeElementNode(n as IRRElement, {
        doc,
        mirror,
        rootId,
      });
    case NodeType.Text: {
      const parentTagName =
        n.parentNode && (n.parentNode as IRRElement).tagName;
      const isStyle = parentTagName === 'STYLE' ? true : undefined;
      return {
        type: NodeType.Text,
        textContent: n.textContent || '',
        isStyle,
        rootId,
      };
    }
    case NodeType.CDATA:
      return {
        type: NodeType.CDATA,
        textContent: '',
        rootId,
      };
    case NodeType.Comment:
      return {
        type: NodeType.Comment,
        textContent: (n as IRRComment).textContent || '',
        rootId,
      };
    default:
      return false;
  }
}

function getRootId(doc: IRRDocument, mirror: Mirror): number | undefined {
  if (!mirror.hasNode(doc)) return undefined;
  const docId = mirror.getId(doc);
  return docId === 1 ? undefined : docId;
}

function getValidTagName(element: IRRElement): string {
  const processedTagName = element.tagName.toLowerCase().trim();
  const tagNameRegex = new RegExp('[^a-z0-9-_:]');
  if (tagNameRegex.test(processedTagName)) {
    // if the tag name is odd and we cannot extract
    // anything from the string, then we return a
    // generic div
    return 'div';
  }

  return processedTagName;
}

function serializeElementNode(
  n: IRRElement,
  options: {
    doc: IRRDocument;
    mirror: Mirror;
    rootId: number | undefined;
  },
): serializedNode | false {
  const { mirror, rootId } = options;
  const tagName = getValidTagName(n);
  const attributes: attributes = n.attributes;
  // media elements
  if (tagName === 'audio' || tagName === 'video') {
    attributes.rr_mediaState = (n as RRMediaElement).paused
      ? 'paused'
      : 'played';
    attributes.rr_mediaCurrentTime = (n as RRMediaElement).currentTime || '';
  } else if (
    (tagName === 'link' || tagName === 'style') &&
    n.textContent !== ''
  ) {
    // the child text is inserted and untracked by the rrweb replayer
    if (n.childNodes[0] && mirror.getId(n.childNodes[0]) < 0)
      attributes._cssText = n.textContent || '';
  } else if (tagName === 'canvas' && (n as RRCanvasElement).rr_dataURL) {
    attributes.rr_dataURL = (n as RRCanvasElement).rr_dataURL as string;
  }
  if (n.scrollLeft) {
    attributes.rr_scrollLeft = n.scrollLeft;
  }
  if (n.scrollTop) {
    attributes.rr_scrollTop = n.scrollTop;
  }

  if (tagName == 'iframe' && !(n as RRIFrameElement).contentDocument) {
    // we can't record it directly as we can't see into it
    // preserve the src attribute so a decision can be taken at replay time
    attributes.rr_src = attributes.src;
  }

  const meta = mirror.getMeta(n);
  const result: serializedNode = {
    type: NodeType.Element,
    tagName,
    attributes,
    childNodes: [],
    isSVG: (meta as elementNode).isSVG,
    rootId,
  };
  if ((meta as elementNode).needBlock !== undefined)
    result.needBlock = (meta as elementNode).needBlock;
  return result;
}

export function serializeNodeWithId(
  n: IRRNode,
  options: {
    doc: IRRDocument;
    mirror: Mirror;
    skipChild: boolean;
    onSerialize?: (n: IRRNode) => unknown;
    onIframeLoad?: (
      iframeNode: RRIFrameElement,
      node: serializedNodeWithId,
    ) => unknown;
    isShadowDom?: boolean;
  },
): serializedNodeWithId | null {
  const {
    doc,
    mirror,
    skipChild = false,
    onSerialize,
    onIframeLoad,
    isShadowDom = false,
  } = options;
  const oldMeta = mirror.getMeta(n);
  // This node doesn't exist in the original web page.
  if (!oldMeta || oldMeta.id < 0) return null;

  const _serializedNode = serializeNode(n, {
    doc,
    mirror,
  });
  if (!_serializedNode) return null;

  const serializedNode = Object.assign(_serializedNode, { id: oldMeta.id });

  if (onSerialize) {
    onSerialize(n);
  }

  const recordChild = !skipChild;

  if (
    (serializedNode.type === NodeType.Document ||
      serializedNode.type === NodeType.Element) &&
    recordChild
  ) {
    const bypassOptions = {
      doc,
      mirror,
      skipChild,
      onSerialize,
      onIframeLoad,
      isShadowDom,
    };
    for (const childN of Array.from(n.childNodes)) {
      const serializedChildNode = serializeNodeWithId(childN, bypassOptions);
      if (serializedChildNode) {
        serializedNode.childNodes.push(serializedChildNode);
      }
    }

    if (n.RRNodeType === NodeType.Element && (n as IRRElement).shadowRoot) {
      serializedNode.isShadowHost = true;
      const shadowRoot = (n as IRRElement).shadowRoot;
      if (shadowRoot) {
        bypassOptions.isShadowDom = true;
        for (const childN of Array.from(shadowRoot.childNodes)) {
          const serializedChildNode = serializeNodeWithId(
            childN,
            bypassOptions,
          );
          if (serializedChildNode) {
            serializedChildNode.isShadow = true;
            serializedNode.childNodes.push(serializedChildNode);
          }
        }
      }
    }
  }

  if (isShadowDom) serializedNode.isShadow = true;

  if (
    serializedNode.type === NodeType.Element &&
    serializedNode.tagName === 'iframe'
  ) {
    const iframeDoc = (n as RRIFrameElement).contentDocument;
    if (iframeDoc) {
      const serializedIframeNode = serializeNodeWithId(iframeDoc, {
        doc: iframeDoc,
        mirror,
        skipChild: false,
        onSerialize,
        onIframeLoad,
      });
      if (serializedIframeNode && onIframeLoad)
        onIframeLoad(n as RRIFrameElement, serializedIframeNode);
    }
  }

  return serializedNode;
}

export function snapshot(
  n: IRRDocument,
  options?: {
    mirror?: Mirror;
    skipChild?: boolean;
    onSerialize?: (n: IRRNode) => unknown;
    onIframeLoad?: (
      iframeNode: RRIFrameElement,
      node: serializedNodeWithId,
    ) => unknown;
  },
): serializedNodeWithId | null {
  const { mirror = new Mirror(), onSerialize, onIframeLoad } = options || {};
  return serializeNodeWithId(n, {
    doc: n,
    mirror,
    skipChild: false,
    onSerialize,
    onIframeLoad,
  });
}

export default snapshot;
