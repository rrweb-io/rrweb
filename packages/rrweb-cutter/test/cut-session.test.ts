/**
 * @jest-environment jsdom
 */
import path from 'path';
import fs from 'fs';
import { createMirror, snapshot } from 'rrweb-snapshot';
import { EventType, SyncReplayer } from 'rrweb';
import { IncrementalSource, eventWithTime, metaEvent } from '@rrweb/types';
import { RRDocument, RRElement, buildFromDom, printRRDom } from 'rrdom';
import { cutSession, getValidSortedPoints } from '../src';
import { snapshot as RRDomSnapshot } from '../src/snapshot';
import { eventsFn as inlineStyleEvents } from './events/inline-style.event';
import mutationEvents from './events/mutation.event';
import inputEvents from './events/input.event';
import multipleSnapshotEvents from './events/multi-fullsnapshot.event';
import scrollEvents from './events/scroll.event';

describe('cut session', () => {
  it('should return the same events if the events length is too short', () => {
    const events1: eventWithTime[] = [];
    const config = { points: [10] };
    expect(cutSession(events1, config)).toEqual([
      {
        events: events1,
        startTimestamp: 0,
        endTimestamp: 0,
        startTime: 0,
        endTime: 0,
      },
    ]);

    const events2: eventWithTime[] = [
      {
        type: EventType.Load,
        data: {},
        timestamp: Date.now(),
      } as eventWithTime,
    ];
    expect(cutSession(events2, config)).toEqual([
      {
        events: events2,
        startTimestamp: events2[0].timestamp,
        endTimestamp: events2[0].timestamp,
        startTime: 0,
        endTime: 0,
      },
    ]);
  });

  it('should return the same events if the points length is 0', () => {
    const events: eventWithTime[] = [
      {
        type: EventType.Load,
        data: {},
        timestamp: Date.now(),
      } as eventWithTime,
      {
        type: EventType.Meta,
        data: {},
        timestamp: Date.now() + 100,
      } as eventWithTime,
    ];
    const config = { points: [] };
    expect(cutSession(events, config)).toEqual([
      {
        events,
        startTimestamp: events[0].timestamp,
        endTimestamp: events[1].timestamp,
        startTime: 0,
        endTime: events[1].timestamp - events[0].timestamp,
      },
    ]);
  });

  it('should sort and validate cutting points array', () => {
    const inputPoints = [10, 250.5, -10, -1, 0, 100];
    expect(getValidSortedPoints([], 100)).toEqual([]);
    expect(getValidSortedPoints(inputPoints, 11)).toEqual([10]);
    expect(getValidSortedPoints(inputPoints, 10)).toEqual([]);
    expect(getValidSortedPoints(inputPoints, 100)).toEqual([10]);
    expect(getValidSortedPoints(inputPoints, 300)).toEqual([10, 100, 250.5]);
  });

  it('should cut the simplest mutation events', () => {
    const events = mutationEvents as eventWithTime[];
    const cutTime1 = 1000,
      cutTime2 = 2000;
    const result = cutSession(events, { points: [cutTime1, cutTime2] });
    expect(result).toHaveLength(3);

    // cut session before 1000ms
    const sessionBefore1s = result[0];
    expect(sessionBefore1s.startTimestamp).toEqual(events[0].timestamp);
    expect(sessionBefore1s.endTimestamp).toEqual(
      events[0].timestamp + cutTime1,
    );
    expect(sessionBefore1s.startTime).toEqual(0);
    expect(sessionBefore1s.endTime).toEqual(cutTime1);
    const cutPoint1Length = 5;
    expect(sessionBefore1s.events).toHaveLength(cutPoint1Length);
    // These events are directly sliced from the original events.
    expect(sessionBefore1s.events).toEqual(events.slice(0, cutPoint1Length));

    // all events between 1000ms and 2000ms
    const sessionBetween1s2s = result[1];
    expect(sessionBetween1s2s.startTimestamp).toEqual(
      events[0].timestamp + cutTime1,
    );
    expect(sessionBetween1s2s.endTimestamp).toEqual(
      events[0].timestamp + cutTime2,
    );
    expect(sessionBetween1s2s.startTime).toEqual(cutTime1);
    expect(sessionBetween1s2s.endTime).toEqual(cutTime2);
    expect(sessionBetween1s2s.events).toHaveLength(3);
    expect(sessionBetween1s2s.events[0].type).toEqual(EventType.Meta);
    expect(sessionBetween1s2s.events[1].type).toEqual(EventType.FullSnapshot);
    expect(sessionBetween1s2s.events[2].type).toEqual(
      EventType.IncrementalSnapshot,
    );
    let replayer = new SyncReplayer(sessionBetween1s2s.events.slice(0, 2)); // only play meta and full snapshot events
    replayer.play();
    // screenshot at 1000ms
    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('screenshot at 1000ms');

    // all events after 2000ms
    const sessionAfter2s = result[2];
    expect(sessionAfter2s.startTimestamp).toEqual(
      events[0].timestamp + cutTime2,
    );
    expect(sessionAfter2s.endTimestamp).toEqual(
      events[events.length - 1].timestamp,
    );
    expect(sessionAfter2s.startTime).toEqual(cutTime2);
    expect(sessionAfter2s.endTime).toEqual(
      events[events.length - 1].timestamp - events[0].timestamp,
    );
    expect(sessionAfter2s.events).toHaveLength(3);
    expect(sessionAfter2s.events[0].type).toEqual(EventType.Meta);
    expect(sessionAfter2s.events[1].type).toEqual(EventType.FullSnapshot);
    expect(sessionAfter2s.events[2].type).toEqual(
      EventType.IncrementalSnapshot,
    );
    replayer = new SyncReplayer(sessionAfter2s.events);
    replayer.play(({ index }) => {
      if (index === 1)
        // full snapshot
        // screen shot at 2000ms
        expect(
          printRRDom(replayer.virtualDom, replayer.getMirror()),
        ).toMatchSnapshot('screenshot at 2000ms');
    });
    // screen shot at 3000ms
    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('screenshot at 3000ms');
  });

  it('should cut events with inline styles', () => {
    const events = inlineStyleEvents() as eventWithTime[];
    const cutTime1 = 1000;
    const result = cutSession(events, { points: [cutTime1] });
    expect(result).toHaveLength(2);
    // all events before 1000ms
    const sessionBefore1s = result[0];
    expect(sessionBefore1s.startTimestamp).toEqual(events[0].timestamp);
    expect(sessionBefore1s.endTimestamp).toEqual(
      events[0].timestamp + cutTime1,
    );
    expect(sessionBefore1s.startTime).toEqual(0);
    expect(sessionBefore1s.endTime).toEqual(cutTime1);
    const cutPoint1Length = 5;
    expect(sessionBefore1s.events).toHaveLength(cutPoint1Length);
    // These events are directly sliced from the original events.
    expect(sessionBefore1s.events).toEqual(events.slice(0, cutPoint1Length));

    // all events after 1000ms
    const sessionAfter1s = result[1];
    expect(sessionAfter1s.startTimestamp).toEqual(
      events[0].timestamp + cutTime1,
    );
    expect(sessionAfter1s.endTimestamp).toEqual(
      events[events.length - 1].timestamp,
    );
    expect(sessionAfter1s.startTime).toEqual(cutTime1);
    expect(sessionAfter1s.endTime).toEqual(
      events[events.length - 1].timestamp - events[0].timestamp,
    );
    expect(sessionAfter1s.events).toHaveLength(3);
    const replayer = new SyncReplayer(sessionAfter1s.events.slice(0, 2)); // only play meta and full snapshot events
    replayer.play();
    // screenshot at 1000ms
    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('screenshot at 1000ms');
  });

  it('new generated session should have a correct meta event', () => {
    const cutTime = 1000;
    const result = cutSession(mutationEvents, { points: [cutTime] });
    expect(result).toHaveLength(2);
    const generatedSession = result[1];
    const originalMetaEvent = mutationEvents.filter(
      (event) => event.type === EventType.Meta,
    )[0] as metaEvent;
    expect(originalMetaEvent.type).toEqual(EventType.Meta);
    const metaEvent = generatedSession.events[0] as eventWithTime;
    expect(metaEvent.type).toEqual(EventType.Meta);
    expect(metaEvent.data).toEqual(originalMetaEvent.data);
    expect(metaEvent.timestamp).toEqual(generatedSession.startTimestamp);
  });

  it('should generate correct meta events from multiple meta events', () => {
    const points = [2000, 4000];
    const results = cutSession(multipleSnapshotEvents, { points });
    expect(results).toHaveLength(3);

    // All meta events in the original events
    const metaEvents = multipleSnapshotEvents.filter(
      (e) => e.type === EventType.Meta,
    );
    expect(metaEvents).toHaveLength(2);

    // The first session should have the first meta event
    expect(
      results[0].events.filter((e) => e.type === EventType.Meta)[0].data,
    ).toEqual(metaEvents[0].data);

    // The session after 2000ms should still have the first meta event
    expect(
      results[1].events.filter((e) => e.type === EventType.Meta)[0].data,
    ).toEqual(metaEvents[0].data);

    // The session after 4000ms should start from the second meta event
    expect(
      results[2].events.filter((e) => e.type === EventType.Meta)[0].data,
    ).toEqual(metaEvents[1].data);
  });

  it('should cut events with input events correctly', () => {
    const points = [1000, 1500];
    const results = cutSession(inputEvents, { points });
    expect(results).toHaveLength(3);

    const Input1NodeId = 6,
      Input2NodeId = 33,
      SelectionNodeId = 26;

    let replayer = new SyncReplayer(results[0].events);
    replayer.play();
    let inputElement1 = replayer.getMirror().getNode(Input1NodeId);
    expect(inputElement1).not.toBeNull();
    expect((inputElement1 as RRElement).getAttribute('value')).toEqual(
      'valueA',
    );

    // The new session start from 1000ms
    replayer = new SyncReplayer(results[1].events);
    replayer.play();
    let incrementalInputEvents = results[1].events.filter(
      (e) =>
        e.type === EventType.IncrementalSnapshot &&
        e.data.source === IncrementalSource.Input,
    );
    expect(incrementalInputEvents).toHaveLength(1);
    // The incremental input event should be played within 5ms
    const inputEvent = incrementalInputEvents[0];
    expect(inputEvent.timestamp - results[1].startTimestamp).toBeLessThan(5);
    inputElement1 = replayer.getMirror().getNode(Input1NodeId);
    expect(inputElement1).not.toBeNull();
    expect((inputElement1 as RRElement).inputData).toEqual(inputEvent.data);
    let selectionElement = replayer.getMirror().getNode(SelectionNodeId);
    expect(selectionElement).not.toBeNull();
    expect(selectionElement?.childNodes).toHaveLength(3);
    let inputElement2 = replayer.getMirror().getNode(Input2NodeId);
    expect(inputElement2).not.toBeNull();
    expect(inputElement2?.nodeType).toBe(document.ELEMENT_NODE);

    // The new session start from 1500ms
    replayer = new SyncReplayer(results[2].events);
    replayer.play();
    // The incremental input event should be played within 5ms
    incrementalInputEvents = results[2].events.filter(
      (e) =>
        e.type === EventType.IncrementalSnapshot &&
        e.data.source === IncrementalSource.Input,
    );
    expect(incrementalInputEvents).toHaveLength(3);
    // All incremental input events should be played within 5ms
    expect(
      incrementalInputEvents.filter(
        (e) => e.timestamp - results[2].startTimestamp < 5,
      ),
    ).toHaveLength(3);
    inputElement1 = replayer.getMirror().getNode(Input1NodeId);
    expect((inputElement1 as RRElement).inputData).toEqual(
      incrementalInputEvents[0].data,
    );
    selectionElement = replayer.getMirror().getNode(SelectionNodeId);
    expect(selectionElement).not.toBeNull();
    expect((selectionElement as RRElement).inputData).toEqual(
      incrementalInputEvents[1].data,
    );
    inputElement2 = replayer.getMirror().getNode(Input2NodeId);
    expect(inputElement2).not.toBeNull();
    expect((inputElement2 as RRElement).inputData).toEqual(
      incrementalInputEvents[2].data,
    );
  });

  it('should cut the session with scroll events correctly', () => {
    const results = cutSession(scrollEvents, { points: [1500] });
    expect(results).toHaveLength(2);

    expect(results[0].events).toHaveLength(5);
    expect(results[1].events).toHaveLength(5);
    const replayer = new SyncReplayer(results[1].events);

    replayer.play();
    expect((replayer.virtualDom as RRDocument).scrollData).toEqual({
      source: IncrementalSource.Scroll,
      id: 1,
      x: 200,
      y: 250,
    });
    expect((replayer.getMirror().getNode(6) as RRElement).scrollData).toEqual({
      source: IncrementalSource.Scroll,
      id: 6,
      x: 100,
      y: 150,
    });
  });
});

describe('Build full snapshot events from RRDom', () => {
  it("should build full snapshot events from RRDom's mirror: main.html", () => {
    document.write(getHtml('main.html'));
    const rrdom = new RRDocument();
    const mirror = createMirror();
    // the full snapshot that is built on jsdom
    const originalSnapshot = snapshot(document, { mirror });

    // Create a RRDom according to the jsdom (real dom).
    buildFromDom(document, mirror, rrdom);

    const newFullSnapshot = RRDomSnapshot(rrdom, {
      mirror: rrdom.mirror,
    });
    expect(newFullSnapshot).toEqual(originalSnapshot);
  });
});

function getHtml(fileName: string) {
  const filePath = path.resolve(__dirname, `./html/${fileName}`);
  return fs.readFileSync(filePath, 'utf8');
}
