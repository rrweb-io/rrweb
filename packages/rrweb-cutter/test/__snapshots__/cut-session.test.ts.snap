// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cut session should cut events with inline styles: screenshot at 1000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 RRText text=\\"\\\\n    \\"
      5 META charset=\\"utf-8\\" 
      6 RRText text=\\"\\\\n    \\"
      7 STYLE 
        -1 RRText text=\\"#root { background: yellow; width: 10px; height: 10px; }\\"
      8 RRText text=\\"\\\\n    \\"
      9 STYLE 
        -1 RRText text=\\".block { width: 20px; height: 20px; background: red; }\\"
      10 RRText text=\\"\\\\n  \\"
    11 RRText text=\\"\\\\n  \\"
    12 BODY 
      13 RRText text=\\"\\\\n    \\"
      14 DIV id=\\"root\\" 
        15 RRText text=\\"      \\\\n    \\"
      16 RRText text=\\"\\\\n    \\"
      17 DIV class=\\"block\\" 
"
`;

exports[`cut session should cut the simplest mutation events: screenshot at 1000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
        34 DIV 
          35 RRText text=\\"1\\"
        33 DIV 
          36 RRText text=\\"2\\"
        32 DIV 
          37 RRText text=\\"3\\"
        31 DIV 
          38 RRText text=\\"4\\"
        29 DIV 
          30 RRText text=\\"5\\"
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;

exports[`cut session should cut the simplest mutation events: screenshot at 2000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
        29 DIV 
          30 RRText text=\\"5\\"
        31 DIV 
          38 RRText text=\\"4\\"
        32 DIV 
          37 RRText text=\\"3\\"
        33 DIV 
          36 RRText text=\\"2\\"
        34 DIV 
          35 RRText text=\\"1\\"
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;

exports[`cut session should cut the simplest mutation events: screenshot at 3000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;
